### PR TITLE
Geochemistry database validator

### LINKED ISSUE #14986: Geochemistry database validator
## Reason
The geochemistry module uses a database of reactions to build the geochemical system. The JSON format that we use is good as jsoncpp throws informative messages if there is a missing brace or comma etc, but we do require the database to fit a certain format for proper working (eg, all logk entries should have the same length as the number of temperatures etc).

To avoid having to check things like vector sizes and values each time we use them, we can check the database follows the required format. This is to catch any errors that may creep in if the database file is edited by hand and provide informative error messages to the user pointing them to the mistake.

## Design
This can be a seperate utility class that is called after reading the database to ensure that it is valid.

## Impact
Avoid mistakes that may be introduced by a user editing the database to add specific reactions that are not currently in the database

### PR DESCRIPTION
**Pull request by cpgr**
Runs once after reading the database to ensure that all entries
meet the required format (size, type, etc).

Testing this was quickly becoming combinatorial! I think I have unit tested all of the functions though.

Closes #14986

### CODE REVIEW

File: modules/geochemistry/src/utils/GeochemicalDatabaseReader.C [hunk 1]
```diff
@@ -27,6 +28,11 @@ GeochemicalDatabaseReader::read(FileName filename)
   // Read the JSON database
   std::ifstream jsondata(filename);
   jsondata >> _root;
+
+  // Validate the JSON database so that we don't have to check array sizes,
+  // check for conversion issues, etc.
+  GeochemicalDatabaseValidator dbv(filename, _root);
+  dbv.validate();
```
Comment (WilkAndy) [inline modules/geochemistry/src/utils/GeochemicalDatabaseReader.C]: Yea, i like validating when reading.  Better than a standalone validator, which is what i thought you were going to do

**Review by WilkAndy — APPROVED** [review]
I've been mucking around the the database a lot. My most common error is forgetting a comma.  I've figured out how to read the stacktrace that is produced, but maybe you can make the error message more readable in that case?

File: modules/geochemistry/src/utils/GeochemicalDatabaseReader.C (hunk 1 above)
Comment (cpgr) [reply modules/geochemistry/src/utils/GeochemicalDatabaseReader.C]: It seemed more useful to do it when running so we can give a useful error.

**Comment by cpgr** [comment]
> I've been mucking around the the database a lot. My most common error is forgetting a comma. I've figured out how to read the stacktrace that is produced, but maybe you can make the error message more readable in that case?

I think that `jsoncpp` throws an error before it would get to this step. Maybe we could catch it and give a better error message like "Looks like you've forgotten a comma again Andy"

**Comment by WilkAndy** [comment]
> > I've been mucking around the the database a lot. My most common error is forgetting a comma. I've figured out how to read the stacktrace that is produced, but maybe you can make the error message more readable in that case?
> 
> I think that `jsoncpp` throws an error before it would get to this step. Maybe we could catch it and give a better error message like "Looks like you've forgotten a comma again Andy"

Yea, i think something along those lines would be great.

**Review by permcody — APPROVED** [review]

**Pull request merged**